STEP 4-1 파일 생성

jgj0131 opened the pull request:
step 4-1 코드 입니다!!

Commit:
STEP 4-1 파일 생성

godrm commented:
우선 (코드리뷰에 앞서서) git commit 로그를 작성하는 방식, commit 하는 작업 단위를 어느정도로 해야 하는지 학습해보세요.
지금처럼 PR에 커밋 하나만 보내면 안됩니다. 커밋은 변경하는 최소 작업단위로 작게 나눠야 합니다.
그래야 원하는 커밋만 선택할 수 있습니다. 이 부분을 정리해보세요.
PR을 숙제검사가 아닙니다. 개발자가 요구사항을 만족하는 코드를 만들기 위해 어떤 작업을 했는지 볼 수 있어야 합니다.

godrm commented on LadderGame/LadderGame/main.swift:
```diff
@@ -1,10 +1,43 @@
+    }else{
+        print("숫자(정수)를 입력하세요")
+    }
+}
```
- make_ladder 함수를 역할 단위로 더 작은 함수로 분리해보세요.
- 스위프트 API 디자인 가이드라인 문서를 읽고 변수, 함수, 상수 이름을 개선해보세요.
- 코드도 글쓰기처럼 공백, 띄어쓰기, 줄간격, 사소한 것들까지도 신경써보세요.

Commit:
STEP4-2

- 변수,함수,상수의 이름을 API 가이드라인을 참고하여 변경하였습니다
- 기존 make_ladder() 함수를 기능별로 나누었습니다
- 띄어쓰기,줄바꿈을 수정하였습니다
- 최소 작업단위로 커밋을 해야하는데 이미 저장이 되어버려 하지 못 했습니다. 다음작업부터는 최소단위별로 커밋하겠습니다

jgj0131 commented:
보내주신 피드백을 바탕으로 코드수정하였습니다

godrm commented:
여전히 커밋로그에 작업한 내용이 포함되어 있지 않네요. 
커밋 로그만 봐도 무슨 작업을 했는지 알 수 있어야 합니다. 어떤 내용을 확인해봤나요? 
위에 명시한 내용을 하나씩 커밋으로 나눠서 여러번 커밋해야 합니다.

jgj0131 commented:
- 기존에 하나의 함수로 구현했었는데 각각 수를 입력받는 함수, 사다리 모양을 2차원 배열로 만드는 함수, 2차원 배열을 출력하는 함수로 나누었습니다
- 함수나 변수명에 _를 다 제거하고 API 가이드라인의 내용을 토대로 이름을 바꾸었습니다
- 공백이나 줄바꿈 같은 경우도 통일 시켰습니다

godrm commented:
다음부터는 위에 적은 항목 하나마다 커밋을 나눠서 해보세요. 아주 작은 단위로 나눠야 한다는 점을 유의하세요.

godrm commented on LadderGame/LadderGame/main.swift:
```diff
@@ -1,43 +1,60 @@
 }
-make_ladder()
+
+outputLadder(makeLadder(inputCount()))
```
여기처럼 함수를 중첩하기 보다는 풀어서 쓰세요. 오히려 더 읽기 어렵습니다.

jgj0131 replied:
한줄이라 오히려 간단할거라 생각했네요 바로 수정하겠습니다

godrm commented on LadderGame/LadderGame/main.swift:
```diff
@@ -1,43 +1,60 @@
+    guard let people = Int(peopleCount!) else { return nil }
+    guard let ladder = Int(ladderCount!) else { return nil }
+    
+    return [people, ladder]
```
`[Int]?` 타입으로 선택한 이유를 설명해주세요.
- 배열을 선택할 때는 순서, 크기, 의도가 명확하게 드러나지 때문에 주의해야 합니다. 
- 배열이 옵셔널인것과 배열이 비어있는 것과 어떤 차이가 있을까요?
- `inputCount()->[Int]?` 라는 함수 시그니처를 보고 어떤 역할을 하는지 명확한 의도를 알 수 있어야 합니다. 입력 카운트를 배열로 리턴하는 함수일까요?

jgj0131 replied:
- [Int]? 타입으로 선택한 이유는 0이 입력되었을때와 문자가 입력되었을때를 나누어서 결과를 보여주고 싶었기 때문이었는데 1이상의 정수가 입력되지 않았을때 에러 메세지를 출력하도록 변경하는 과정에서 따로 수정을 하지 않았기 때문입니다
- 배열이 옵셔널인것은 메모리 할당이 안 되어있는 상태이고, 비어있는것은 값은 없지만 메모리는 할당되어있는 차이가 있다고 생각합니다!
- 만약 inputCount()에 출력값이 nil이라면 배열로 리턴하지 않을것같습니다

godrm commented on LadderGame/LadderGame/main.swift:
```diff
@@ -1,43 +1,60 @@
+
+/// 사람의 숫자와 사다리의 숫자를 입력 받아 사다리를 만들고 출력하는 함수
+func makeLadder(_ inputCount : [Int]?)->[[String]] {
+    if inputCount != nil && inputCount![0] > 0 && inputCount![1] > 0{
```
매개변수를 배열로 처리하고 나니 복잡한 비교식이 많아졌네요. 이런 경우는 배열을 사용하지 않고 개별 변수를 사용하는게 더 명확할 것 같습니다.
심지어 아래에서 people, ladder 로 다시 할당하고 있다는 것은 매개변수가 잘못 선언됐다는 걸로 생각됩니다.

jgj0131 replied:
makeLadder()의 인자를 배열에서 countOfPeople,countOfLadder를 인자로 넣기 위해 inputCount()의 반환을 [Int]?에서 두개의 변수(Int)로 반환하도록 수정했고, makeLadder()도 수정하였습니다

godrm reviewed (commented):
아직 의도를 드러내서 코드에 표현하는 데 부족한 것 같습니다.
변수, 상수 이름에 Array, Int, String, Num 같은 타입에 대한 표현보다는 어떤 역할인지 의도를 나타내는 단어를 쓰는게 좋습니다. 다음 피드백을 확인해보세요.

godrm commented:
아까는 한꺼번에 댓글을 남겼는데, 이제는 제가 남긴 코멘트별로 어떤 내용을 수정했는지 댓글을 적어주면 확인하고 넘어가겠습니다.

Commit:
inputCount() 함수 수정

Commit:
makeLadder() 함수 수정

godrm commented:
커밋 로그 작성하는 방법에 대해 학습하고 적용해보세요. 
`makeLadder() 함수 수정` 보다 더 구체적으로 무엇을 바꿨는지 작업한 내용을 적어보세요.

Commit:
Modify : 가독성을 위한 함수실행부 수정

godrm commented on LadderGame/LadderGame/main.swift:
```diff
@@ -1,58 +1,58 @@
     } else {
-        for i in ladderArray {
+        for i in ladders {
             for j in i{
```
for 문에서 사용하는 i, j 변수 이름도 의도하는 역할에 대해 명확하게 작성하세요.

jgj0131 replied:
넵! 감사합니다!

jgj0131 replied:
for문 내의 I,j를 각각 ladderIndex, peopleIndex로 수정하였습니다

Commit:
Modify : 가독성을 위한 for문 내의 변수명 수정

godrm commented:
수고하셨습니다. 다음 단계를 진행하면서 오늘 주고받았던 내용도 함께 고민해서 진행해보세요.
